legacy.coffee

```diff
@@ -0,0 +1,291 @@
+		request.getAsync("#{@protocol}://#{@registry}:#{@port}#{path}")
+
+	head: (path) ->
+		request.headAsync("#{@protocol}://#{@registry}:#{@port}#{path}")
+
+	# Convert to string in the format registry.tld:port
+	toString: ->
+		return "#{@registry}:#{@port}"
+
+class RegistryV1 extends Registry
+	getVersion: -> 1
+
+	# Get the id of an image on a given registry and tag.
+	getImageId: (imageName, tagName) ->
+		@get("/v1/repositories/#{imageName}/tags")
+		.spread (res, data) =>
+			if res.statusCode == 404
+				throw new Error("No such image #{imageName} on registry #{@registry}")
+			if res.statusCode >= 400
+				throw new Error("Failed to get image tags of #{imageName} from #{@registry}. Status code: #{res.statusCode}")
+			tags = JSON.parse(data)
+			if !tags[tagName]?
+				throw new Error("Could not find tag #{tagName} for image #{imageName}")
+			return tags[tagName]
+
+	# Return the ids of the layers of an image.
+	getImageHistory: (imageId) ->
+		@get("/v1/images/#{imageId}/ancestry")
+		.spread (res, data) =>
+			if res.statusCode >= 400
+				throw new Error("Failed to get image ancestry of #{imageId} from #{@registry}. Status code: #{res.statusCode}")
+			history = JSON.parse(data)
+			return history
+
+	# Return the number of bytes docker has to download to pull this image (or layer).
+	getImageDownloadSize: (imageId) ->
+		@get("/v1/images/#{imageId}/json")
+		.spread (res, data) =>
+			if res.statusCode >= 400
+				throw new Error("Failed to get image download size of #{imageId} from #{@registry}. Status code: #{res.statusCode}")
+			return parseInt(res.headers['x-docker-size'])
+
+	getLayerDownloadSizes: (imageName, tagName) ->
+		layerSizes = {}
+		layerIds = []
+		@getImageId(imageName, tagName)
+		.then (imageId) =>
+			@getImageHistory(imageId)
+		.map (layerId) =>
+			layerIds.push(layerId)
+			@getImageDownloadSize(layerId)
+			.then (size) ->
+				layerSizes[layerId] = size
+		.return([ layerSizes, layerIds ])
+
+class RegistryV2 extends Registry
+	getVersion: -> 2
+
+	# Return the ids of the layers of an image.
+	getImageLayers: (imageName, tagName) ->
+		@get("/v2/#{imageName}/manifests/#{tagName}")
+		.spread (res, data) =>
+			if res.statusCode >= 400
+				throw new Error("Failed to get image ancestry of #{imageName} from #{@registry}. Status code: #{res.statusCode}")
+			_.map(JSON.parse(data).fsLayers, 'blobSum')
+
+	# Return the number of bytes docker has to download to pull this blob.
+	getLayerDownloadSize: (imageName, blobId) ->
+		@head("/v2/#{imageName}/blobs/#{blobId}")
+		.spread (res, data) =>
+			if res.statusCode >= 400
+				throw new Error("Failed to get image download size of #{imageName} from #{@registry}. Status code: #{res.statusCode}")
+			return parseInt(res.headers['content-length'])
+
+	# Gives download size per layer (blob)
+	getLayerDownloadSizes: (imageName, tagName) ->
+		layerSizes = {}
+		layerIds = []
+		@getImageLayers(imageName, tagName)
+		.map (blobId) =>
+			layerIds.unshift(blobId)
```

Comment:
You can use
```coffee
.then (layerIds) ->
	layerSizes = {}
	Promise.map layerIds, (layerId) ->
		@getLayerDownloadSize(layerId)
		.then (size) ->
			layerSizes[layerId] = size
	.return([ layerSizes, layerIds.reverse() ])
```
to reverse the array we already have, without needing any `unshift` or `push` at all.  Which I think makes sense to do in order to avoid building up a new array and also to make it clear that we are intentionally reversing the array - and the same will work for v1 to avoid recreating the array

And with the array available it's also possible to use lodash to handle the object creation with
```coffee
.then (layerIds) ->
	layerSizes =_.keyBy(layerIds).mapValues (blobId) =>
		@getLayerDownloadSize(imageName, blobId)
	Promise.all([ layerIds, Promise.props(layerSizes) ])
```
or
```coffee
.then (layerIds) ->
	Promise.props _.keyBy(layerIds).mapValues (blobId) =>
		@getLayerDownloadSize(imageName, blobId)
	.then (layerSizes) ->
		[ layerIds, layerSizes ]
```